### PR TITLE
Add CDK support

### LINKED ISSUE #48: AWS CDK Support
## Feature Description

Instead of creating the basti resources with the command `basti init` there would also be an option of creating the resources through an AWS CDK construct. 

An AWS CDK construct for Basti would offer the following benefits:

* Easier to assign permissions: AWS CDK would allow users to easily assign permissions to Basti bastion hosts. This would make it easier to ensure that Basti bastion hosts only have the permissions that they need.
* Easier to set up monitoring: AWS CDK would allow users to easily set up alarms and other monitoring for Basti bastion hosts. This would help users to keep Basti bastion hosts secure and healthy.
* More consistent infrastructure: AWS CDK would allow users to create a consistent infrastructure for their Basti bastion hosts. This would make it easier for users to manage and maintain their Basti bastion hosts.
* Infrastructure is better documented: AWS CDK is infrastructure-as-code, a construct would mean that Basti resources are also documented, this makes it more transparent to maintain. 

## Use Case

Currently, users can create Basti bastion hosts using the basti init command. However, this can be difficult to use when all other resources are manged through AWS CDK, as it requires users to manually assign permissions and set up monitoring. An AWS CDK construct would make it easier for these users to create Basti bastion hosts, as it would allow them to use the AWS CDK's built-in permission and monitoring features.

## Proposed Solution

The construct should do a lot, but should not be too configurable. The interface should be relatively simple. As targets are most often selected when connecting to the bastion host, the initial version should just set up the bastion host and no more. The interface in python (CDK supports many languages through JSII) would just be a class.
```
from basti_cdk import BastiHost

host = BastiHost(self, 'bastiHost')
```

The basti CDK construct can then create its own VPC and any subnets it needs. Maybe the interface can be expanded that these are created.

The construct could also be more limiting and only allow access to a specific database/database cluster. 

## Related Issues/PRs

No related PRs

## Notes
I would be able to help in creating this functionality.

### PR DESCRIPTION
**Pull request by bobveringa**
## Proposed Changes

This PR adds initial support for CDK to Basti. It is currently still a draft for gathering feedback.

## Related Issue

[Closes #48]  

## Changes

The `.eslintrc.json` file was modified because WebStorm as going crazy about line endings and this made it silent.

It contains a `README.md` but only as template.

Users have the option of explicitly setting the ID of the basti instance. If the user does not provide one, it generates one instead based on the sha1 hash of the construct ID. This is "random enough" for the usecase.

The basti instance itself is also configurable with a custom machine image and instance type. This could be expanded more in the future. 

The subnet is selected automatically to whatever Public subnet cloudformation things it can use. However, users do have the option to pass a specific `SubnetSelection` object so this can be modified if needed.

The `BastiInstance` also has an additional method called `grantBastiCliConnect` this function takes `IGrantable` as a parameter and allows developers to grant roles (and IAM users) permissions to connect to the basti instance using the basti CLI. In the future, the construct might support other methods that would allow other services to connect. But for now, this will suffice.

### CODE REVIEW

**Comment by bobveringa** [comment]
@BohdanPetryshyn If you could provide some feedback on this while I continue to work on setting up tests, that would be much appreciated.

**Comment by BohdanPetryshyn** [comment]
@bobveringa, I will be able to review the PR tomorrow morning. Thank you!

**Comment by bobveringa** [comment]
A couple of things that still need to be figured out. Setting the author correctly so that it can be published to PyPI. We also need to determine who is the official maintainer of this package. And maybe we have to set up GitHub actions for basti in general?

File: packages/basti-cdk/src/basti-instance.ts
Comment (BohdanPetryshyn) [inline packages/basti-cdk/src/basti-instance.ts:168]: This method will be super-useful 👍 👍 👍

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [inline packages/basti-cdk/src/basti-access-security-group.ts:29]: Users might think they have to pass the bastion instance ID here. Let's add a line to the documentation.

**Comment by BohdanPetryshyn** [comment]
> maybe we have to set up GitHub actions for basti in general
100% It was sometimes hard for me to publish even a single Basti CLI package manually

> who is the official maintainer of this package
I thought I will be maintaining everything unless you are willing to help with this 🤗

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:60]: The reason I set it up like this is that sometimes these constructs take many arguments. Having everything top level makes it a lot more difficult to see which parameters are related to each other.

I can flatten it out, but I would prefer not too. Instead, I can add documentation as to what a `machineConfig` is. I think that it can have a strong definition, something similar to.

machineConfig is all the parameters that influence the underlying EC2 instance. This could be from block devices that should be attached or the size of current storage.

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:149]: I checked and it does not, should it?

File: packages/basti-cdk/src/bastion-cloudinit.ts
Comment (bobveringa) [reply packages/basti-cdk/src/bastion-cloudinit.ts:7]: Maybe this would also be the time to rename the `basti` directory to `basti-cli` to avoid confusion?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-access-security-group.ts:60]: I'm not entirely sure what you mean by this. Could you please explain a bit more what you mean by lookup?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-access-security-group.ts:25]: The port is currently specified in the `addBastiInstance` method. While I don't you need to specify this per instance you are adding, it is probably the easiest/cleanest way to do this. As you define the access security group before the bastion host, which means you might not know the port yet (depending on how your database is set up).

This way, you can get the port from the rds instance (or cluster) when allowing basti access.

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:168]: 😄

**Comment by bobveringa** [comment]
Setting up actions might be slightly more complicated with the monorepo structure, but I doubt basti is the only package using that approach, so there must be some solution.

Yeah, it makes sense to have you as maintainer for everything. In that case, you might need to set up a PyPI account, I'll investigate this and make sure this can be done swiftly.

**Comment by BohdanPetryshyn** [comment]
Thank you @bobveringa! I have a very busy day today so I'll only be able to reply to your comments in the evening or tomorrow morning

File: packages/basti-cdk/src/basti-instance.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-instance.ts:149]: I think it should as it might be required by a Tagging Policy ([related comment](https://github.com/BohdanPetryshyn/basti/issues/50#issuecomment-1636734387)) I'm currently working on tags support in `basti init` and I added tagging for all the related resources. Currently, we only create an EBS volume and an ENI.

@DrFunk-n-stein please, correct me if I'm wrong

Comment (DrFunk-n-stein) [reply packages/basti-cdk/src/basti-instance.ts:149]: Yup, that's right. 

If tagging is enforced by a combination of tagging policy and service control policy, the creation of basti would simply be denied if the instance is not created with tags applied. 
So adding them afterwards is not an option here unfortunately.

Comment (DrFunk-n-stein) [reply packages/basti-cdk/src/basti-instance.ts:149]: See [here](https://aws.amazon.com/blogs/mt/implement-aws-resource-tagging-strategy-using-aws-tag-policies-and-service-control-policies-scps/)

**Comment by bobveringa** [comment]
That's ok, I'm, not in a hurry.

I also noticed that the package-lock.json should probably be removed from the basti package. https://docs.npmjs.com/cli/v9/using-npm/workspaces

**Comment by bobveringa** [comment]
> Yeah, it makes sense to have you as maintainer for everything. In that case, you might need to set up a PyPI account, I'll investigate this and make sure this can be done swiftly.

Looks like there is a package by AWS for this `jsii-release` https://www.npmjs.com/package/jsii-release after the package is finalized we can go over how to set this up if you'd like.

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:63]: @bobveringa, I meant something like the [VPC.fromLookup](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.Vpc.html#static-fromwbrlookupscope-id-options) method. Such a method will be useful when the `BastiInstance` is defined in another CDK application and the user only has its `bastiId`

Comment (bobveringa) [reply packages/basti-cdk/src/basti-access-security-group.ts:63]: Ah yeah, that is a good idea. But it might be a bit difficult to implement. The BastiInstance construct creates 3 resources:
1. ec2.Instance
2. iam.Role
3. ec2.SecurityGroup

The resources `iam.Role` and `ec2.SecurityGroup` both allow the creation of their interface counterparts (`ISecurityGroup` and `IRole`) using the `.fromName` (or similarly named) method. However, it seems like `Instance` does not support any `.form...` methods. There have been requests for this (see https://github.com/aws/aws-cdk/issues/3911) but it seems like AWS is not interested in supporting it.

That leaves us with a few options:
1. We support it partially, we only look up both the security group and role from the ID but do not import the instance. This feels a bit wrong, but it would support this use case. 
2. Fully implement the `instance.fromLookup()` method and get it pushed in AWS. This would likely make more people happy, but there is probably a reason they didn't implement this, which means there must be something complex.
3. Wait for CDK support? This would likely take a while, since that issue I linked is from 2019.
4. Use a workaround.There is a workaround available (it is not great, but it is there):
All that the `bastiInstance` is used for here is the security group it contains. And given that `BastiAccessSecurityGroup` is essentially just a regular `SecurityGroup` but with some fluff. Users could export the name/id of the `basti.securityGroup` and just add an ingress line for that group at the given port. It is not as neat but achieves the same purpose.

None of these options are particularly great, 1. Feels not great because if we implement something like `BastiInstance.fromId()` and it then doesn't provide the basti instance, then that's no good. 2 and 3 are unlikely to happen (unless someone is brave enough to do them). And 4 is a workaround that is not that pretty.

If you feel like that, this functionality is vital for release than we can certainly implement this. But it doesn't add that much functionality, and there is a workaround that is available for people that have set-ups like that. If people really need this functionality we can revisit it.

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:149]: @DrFunk-n-stein From all the resources that are created by the construct
* SecurityGroup
* Role
* Instance

Which of those need tags, only the Instance?

As for the instance itself, I'm not entirely sure how to handle this in CDK. The ec2.Instance construct does have the `propagateTagsToVolumeOnCreation` option. How do you normally create these instances, is there an example I can use?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:28]: I agree here, this is the best solution that comes to my mind as well.

File: .github/workflows/basti-cdk-actions.yml
Comment (BohdanPetryshyn) [inline .github/workflows/basti-cdk-actions.yml:27]: npm -w flag should already do this for us

**Review by BohdanPetryshyn — COMMENTED** [review]
Thank you for such a comprehensive approach to contribution (docs and the GH action) 👍 

What about scheduling a call for a quick "demo"? It would be awesome to see the construct in action. Also, I will be able to better understand the building a publishing process, ask some more questions, and provide more feedback. Let's agree on a call on Linkedin.

I found some typos in the docs. But in general, I plan to improve consistency and apply other style-related minor changes in docs later on my own. This will be much more efficient than the review flow.

**Comment by BohdanPetryshyn** [comment]
> I also noticed that the package-lock.json should probably be removed from the basti package.

You're right. Could you please, do this in your PR?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:63]: I think the first option is not bad at all 👍 

We could add a limited `IBastiInstance` interface and a corresponding implementation that would be returned from the `BastiInstance.fromBastiId` method. This limited implementation will be usable in the `BastiAccessSecurityGroup. addBastiInstance` method and maybe in other use cases that we haven't thought of as well.

Later, as AWS adds the `from...()` method for EC2 instances, we could expand the `IBastiInstance` interface.

I think it would be great to have it right from the first version if it is not too hard to implement.

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:149]: Tags are now added to all resources that are created, This might be overkill, but should work.

File: packages/basti-cdk/package.json
Comment (bobveringa) [reply packages/basti-cdk/package.json:6]: I had this configured as `dist/index.js` but when exporting to python using JSII it wouldn't work. I also don't think it is the output folder. 
>  The main field is a module ID that is the primary entry point to your program. ...

[docs](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#main)

I already tried a bunch of options to get this to dist/index.js but nothing I did made it work when exporting to python. Which then just complained. No such file dist/....

Comment (BohdanPetryshyn) [reply packages/basti-cdk/package.json:6]: We'll cover this on a call 👌

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [inline packages/basti-cdk/src/basti-access-security-group.ts:62]: What about renaming this method to `allowConnection`? I think it's more obvious what it does this way.

**Review by BohdanPetryshyn — APPROVED** [review]

File: packages/basti-cdk/src/basti-instance.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-instance.ts:158]: Just to confirm, tags are added to EBS Volumes and ENIs as well?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (BohdanPetryshyn) [inline packages/basti-cdk/src/basti-access-security-group.ts:12]: `BastiAccessSecurityGroupProps` should accept tags as well

Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: Hmm 🤔 Maybe we don't need to manage custom user-defined tags on our own and rely on the `Tags.of` method instead?

What I mean is that users will be able to add their tags this way:

```js
const bastiInstance = new BastiInstance(/*...*/);

Tags.of(bastiInstance).add('hello', 'world');
```

Looks like this is the unified approach in CDK. Let's see what other popular CDK libraries do

Comment (DrFunk-n-stein) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: This is fine from my point of view. 

Stack tags are propagated to all taggable resources

Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: Thank you for your input @DrFunk-n-stein!

Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: @DrFunk-n-stein, I would really appreciate it if you could share some overall feedback regarding the construct interface we came up with. I'm sure you have much more experience in CF/CDK than I do 😊

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:158]: Tags are propagated to the EBS volume. I'm not sure about the ENI, all the things that I could assign tags to now have tags. But I have no idea how CDK/CloudFormation manage these changes. Even if it didn't assign tags to the ENI there is no way (that I know of) to add more tags to the instance.

The `BastiInstance` Construct exposes all resources it creates, so users would be able to add tags in the same way the basti construct can. That should cover all the instances where the Basti construct somehow does not yet assign tags. People that have these more restrictive tagging policies can then contribute their changes back to Basti, and we then integrate them.

Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:158]: Also if I understand @DrFunk-n-stein correctly, it is not needed to apply tags to every single resource in CDK, as you can assign Tags to a stack, and they are then propagated to all taggable resources?

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: ```Tags.of(bastiInstance).add('hello', 'world');```

Should work like any other construct since we are just extending the `SecurityGroup`

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-access-security-group.ts:62]: Since we are extending the `SecurityGroup` construct, it might make sense to name it something with `basti` in it. Just to it is clear what methods basti is adding. Something like `allowBastiConnection` or `bastiAllowConnection`

Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-access-security-group.ts:62]: Makes sense 👍 

What about `allowConnectionFromBastiInstnace` or `allowBastiInstanceConnection`? Kinda long but exactly represents what it does and has `basti` in it

File: packages/basti-cdk/src/basti-instance.ts
Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-instance.ts:158]: I think we can remove the `tags` prop altogether according to https://github.com/BohdanPetryshyn/basti/pull/53#discussion_r1270273126

**Comment by BohdanPetryshyn** [comment]
@bobveringa, we have only two unresolved conversations here. We're good to go after resolving these 🎉

File: packages/basti-cdk/src/basti-instance.ts
Comment (bobveringa) [reply packages/basti-cdk/src/basti-instance.ts:158]: I'll rework it, so it is just for the instance then. We have a use case where it is nice just to be able to tag the instance for something. But for the other items it is not relevant. Would that be ok?

Comment (BohdanPetryshyn) [reply packages/basti-cdk/src/basti-instance.ts:163]: I think this is doable with the `includeResourceTypes` option of the [Tags.of](https://docs.aws.amazon.com/cdk/v2/guide/tagging.html) method. 

I'd suggest implementing advanced tagging separately after some research on how other popular libraries do that. If we add the property right now without carefully thinking, we'll need to maintain or deprecate it later

File: packages/basti-cdk/src/basti-access-security-group.ts
Comment (DrFunk-n-stein) [reply packages/basti-cdk/src/basti-access-security-group.ts:12]: I'm a terraform guy 😅 started with CDK some weeks ago, so please forgive me that I can't give input right now. I think you got it covered

@BohdanPetryshyn

**Pull request merged**
